### ISSUE TITLE
bug in replace array in json

### ISSUE THREAD

**Issue by mohamadk** [flipkart-incubator/zjsonpatch#18]
hi thanks for this great library.
when I'm getting differnt between this two jsons  
"{\"b\": [1,2,4,9]}"
"{\"b\": [6,2,4,9]}"
the difference value currectly return operation replace in location 0.

but if I get diff this two jsons 

     {
     "sections": {

    "array":[
      2,
      2,
      5,
      6
    ],

    "general": {
      "title": "\u0627\u0637\u0644\u0627\u0639\u0627\u062a \u06a9\u0644\u06cc",
      "web_icon": "fa-globe",
      "icon": "fa_globe",
      "view": "general",
      "order": 10,
      "desktop_show": true
    },
    "processes": {
      "title": "\u0641\u0631\u0622\u06cc\u0646\u062f\u0647\u0627",
      "web_icon": "fa-gears",
      "icon": "fa_gears",
      "view": "default",
      "order": 2,
      "desktop_show": true
    },
    "fuel_consumers": {
      "title": "\u062a\u062c\u0647\u06cc\u0632\u0627\u062a",
      "web_icon": "fa-fire",
      "icon": "fa_fire",
      "view": "countable",
      "order": 2,
      "desktop_show": true
    },
    "fuel_tanks": {
      "title": "\u0645\u062e\u0627\u0632\u0646",
      "web_icon": "fa-database",
      "icon": "fa_database",
      "view": "fuel_tanks",
      "order": 4,
      "desktop_show": true
    },
    "vehicles": {
      "title": "\u0645\u0627\u0634\u06cc\u0646\u200c\u0622\u0644\u0627\u062a",
      "web_icon": "fa-truck",
      "icon": "fa_truck",
      "view": "countable",
      "order": 5,
      "desktop_show": true
    },
    "checkpoints": {
      "title": "\u0645\u0648\u0642\u0639\u06cc\u062a\u200c\u0647\u0627\u06cc \u0645\u06a9\u0627\u0646\u06cc",
      "web_icon": "fa-map-marker",
      "icon": "fa_map_marker",
      "view": "default",
      "order": 6,
      "desktop_show": false,
      "children": {
        "aaaa": {
          "title": "\u0627\u0637\u0644\u0627\u0639\u0627\u062a \u06a9\u0644\u06cc",
          "order": 10,
          "desktop_show": true
        },
        "bbb": {
          "order": 2,
          "desktop_show": true
        },
        "ccc": {
          "order": 5,
          "desktop_show": true
        }
      }
    }
      }
     }   
`
and         
 
      {
      "sections": {

    "array":[
      6,
      2,
      5,
      6
    ],

    "general": {
      "title": "\u0627\u0637\u0644\u0627\u0639\u0627\u062a \u06a9\u0644\u06cc",
      "web_icon": "fa-globe",
      "icon": "fa_globe",
      "view": "general",
      "order": 10,
      "desktop_show": true
    },
    "processes": {
      "title": "\u0641\u0631\u0622\u06cc\u0646\u062f\u0647\u0627",
      "web_icon": "fa-gears",
      "icon": "fa_gears",
      "view": "default",
      "order": 2,
      "desktop_show": true
    },
    "fuel_consumers": {
      "title": "\u062a\u062c\u0647\u06cc\u0632\u0627\u062a",
      "web_icon": "fa-fire",
      "icon": "fa_fire",
      "view": "countable",
      "order": 2,
      "desktop_show": true
    },
    "fuel_tanks": {
      "title": "\u0645\u062e\u0627\u0632\u0646",
      "web_icon": "fa-database",
      "icon": "fa_database",
      "view": "fuel_tanks",
      "order": 4,
      "desktop_show": true
    },
    "vehicles": {
      "title": "\u0645\u0627\u0634\u06cc\u0646\u200c\u0622\u0644\u0627\u062a",
      "web_icon": "fa-truck",
      "icon": "fa_truck",
      "view": "countable",
      "order": 5,
      "desktop_show": true
    },
    "checkpoints": {
      "title": "\u0645\u0648\u0642\u0639\u06cc\u062a\u200c\u0647\u0627\u06cc \u0645\u06a9\u0627\u0646\u06cc",
      "web_icon": "fa-map-marker",
      "icon": "fa_map_marker",
      "view": "default",
      "order": 6,
      "desktop_show": false,
      "children": {
        "aaaa": {
          "title": "\u0627\u0637\u0644\u0627\u0639\u0627\u062a \u06a9\u0644\u06cc",
          "order": 10,
          "desktop_show": true
        },
        "bbb": {
          "order": 2,
          "desktop_show": true
        },
        "ccc": {
          "order": 5,
          "desktop_show": true
        }
      }
      }
      }
     }
it return add and remove operation in 0 and 2 positions.
i think this is a bug.

**Comment by vishwakarma** [comment]
Agreed, thanks for pointing out the issue. Clearly, the diffs generated are not optimized in this case.
I will fix this once i  get some space to work on this project, Moreover, i would be glad to accept PR for same.

**Comment by nebevay** [comment]
Hi, I've looked at the code a bit and I think the underlying issue is that REPLACE operations are not collapsed. Take the following arrays:

source: [1, 2, 3, 4, 5]
target:  [5, 1, 4, 3, 2]
lcs:       [1, 2] ([1, 4] or [1, 3] would also be valid)

You will get, REMOVE 1, REPLACE 2 5, ADD 1, REPLACE 3, 4, REPLACE 4, 3, REMOVE 5, ADD 2

To reduce this into moves you would need the following:
REMOVE 1, ADD 1 -> MOVE 1
REMOVE 5, REPLACE 2 5, ADD, 2 -> MOVE 5
REPLACE 3, 4, REPLACE 4, 3 -> MOVE 3

Obviously you need to keep track of coordinates as well. The optimization that is currently in place only collapses REMOVE and ADD into MOVE, it doesn't take into account these chains of replacements.

Sadly I won't have enough time to code this myself right now, but thought it might help if someone else wanted to tackle it.

**Comment by aantoniuk-gl** [comment]
What about this bug? It's been created more than 3 years ago
